### DISCUSSION TITLE
Resources with serverside apply unable to sync managedFields must be nil

### DISCUSSION THREAD

**Discussion by j00p34** [argoproj/argo-cd#20749] [Q&A]
We use sever-side apply on multiple resources but the sync fails with: metadata.managedFields must be nil
We use this among others things to patch the default namespace with an extra label. 
When we manually reset the managedfields to an empty array the sync succeeds one time.
After the sync the managedfields are back again and arogcd is not able to sync the same resource again
This is the patch we are trying to apply

apiVersion: v1
kind: Namespace
metadata:
  name: default
  annotations:
    argocd.argoproj.io/sync-options: ServerSideApply=true, Validate=false
  labels:
    network.openshift.io/policy-group: policygroup
    
 Are we missing something here?

**Comment by andrii-korotkov-verkada** [comment]
Do you or something else set managedFields in the target manifests?

**Comment by j00p34** [comment]
thanks for your reply,

No we are not setting them, here are the managedfields after the first sync for the default namespace, seems like this is generated by k8s or argo
```
argocd or k8s :
managedFields:
    - apiVersion: v1
      fieldsType: FieldsV1
      fieldsV1:
        f:metadata:
          f:annotations:
            f:argocd.argoproj.io/sync-options: {}
            f:openshift.io/sa.scc.mcs: {}
            f:openshift.io/sa.scc.supplemental-groups: {}
            f:openshift.io/sa.scc.uid-range: {}
          f:labels:
            f:app.kubernetes.io/[redacted]: {}
            f:kubernetes.io/metadata.name: {}
            f:network.openshift.io/policy-group: {}
            f:pod-security.kubernetes.io/audit: {}
            f:pod-security.kubernetes.io/enforce: {}
            f:pod-security.kubernetes.io/warn: {}
        f:spec:
          f:finalizers: {}
      manager: argocd-controller
      operation: Apply
    - apiVersion: v1
      fieldsType: FieldsV1
      fieldsV1:
        f:metadata:
          f:annotations:
            .: {}
            f:argocd.argoproj.io/sync-options: {}
            f:openshift.io/sa.scc.mcs: {}
            f:openshift.io/sa.scc.supplemental-groups: {}
            f:openshift.io/sa.scc.uid-range: {}
          f:labels:
            .: {}
            f:app.kubernetes.io/[redacted] {}
            f:kubernetes.io/metadata.name: {}
            f:network.openshift.io/policy-group: {}
            f:pod-security.kubernetes.io/audit: {}
            f:pod-security.kubernetes.io/enforce: {}
            f:pod-security.kubernetes.io/warn: {}
        f:spec:
          f:finalizers: {}
      manager: before-first-apply
      operation: Update
      time: '2024-11-08T15:00:58Z'``
```

**Comment by andrii-korotkov-verkada** [comment]
Ah, interesting, why is `argocd.argoproj.io/sync-options` among managed fields?

**Reply by j00p34** [reply]
No Idea, maybe it's because we are setting server-side apply?
```
metadata: 
  annotations:
    argocd.argoproj.io/sync-options: ServerSideApply=true, Validate=false
```

**Reply by andrii-korotkov-verkada** [reply]
Why are there two managers for roughly the same field? What does `before-first-apply` manager do?

**Reply by j00p34** [reply]
I'm not sure, but I think that's there because I cleared the managedfields with a patch so the sync would work. It seems like k8s is setting this

**Comment by Mutdogus** [comment]
Opened https://github.com/argoproj/argo-cd/pull/28585 to fix this.

Root cause: gitops-engine writes the object to the apply manifest verbatim in `createManifestFile`, and the **apply** path never stripped `metadata.managedFields` (the server-side **diff** path already does, in `pkg/diff/diff.go`). A server-side apply whose body carries `managedFields` is rejected by the API server with exactly `metadata.managedFields must be nil` — which is why clearing them by hand works for a single sync before they repopulate.

I confirmed against a live cluster (kind, k8s v1.36) that a manifest carrying `managedFields` is accepted by CREATE/REPLACE/client-side apply and rejected **only** by server-side apply. The PR strips `managedFields` at that single serialization point, with a unit test. Feedback welcome.
